airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/event/CatalogValidator.kt

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.core.event
+
+import io.airbyte.cdk.core.config.AirbyteConfiguredCatalog
+import io.airbyte.cdk.core.context.env.ConnectorConfigurationPropertySource
+import io.airbyte.cdk.core.operation.OperationType
+import io.airbyte.protocol.models.v0.ConfiguredAirbyteCatalog
+import io.github.oshai.kotlinlogging.KotlinLogging
+import io.micronaut.context.annotation.Requires
+import io.micronaut.context.annotation.Value
+import io.micronaut.context.event.ApplicationEventListener
+import io.micronaut.context.event.StartupEvent
+import jakarta.inject.Singleton
+
+private val logger = KotlinLogging.logger {}
+
+/**
+ * Event listener that validates the Airbyte configured catalog, if present. This listener is
+ * executed on application start.
+ */
+@Singleton
+@Requires(bean = AirbyteConfiguredCatalog::class)
+@Requires(property = ConnectorConfigurationPropertySource.CONNECTOR_OPERATION)
+class CatalogValidator(
+    @Value("\${micronaut.application.name}") private val connectorName: String,
+    @Value("\${airbyte.connector.operation}") private val operation: String,
+    private val airbyteConfiguredCatalog: AirbyteConfiguredCatalog,
+) : ApplicationEventListener<StartupEvent> {
+    companion object {
+        var emptyCatalog: ConfiguredAirbyteCatalog = ConfiguredAirbyteCatalog()
+    }
+
+    override fun onApplicationEvent(event: StartupEvent) {
+        if (requiresCatalog()) {
+            if (emptyCatalog != airbyteConfiguredCatalog.getConfiguredCatalog()) {
+                logger.info { "$connectorName connector configured catalog is valid." }
+            } else {
+                throw IllegalArgumentException("Configured catalog is not valid.")
+            }
+        }
+    }
+
+    private fun requiresCatalog(): Boolean {
+        return OperationType.READ.name.equals(
+            operation,
+            ignoreCase = true,
+        ) || OperationType.WRITE.name.equals(operation, ignoreCase = true)
```

Comment:
Please, no. There has to be a better way. I get that `operation: String` for DI purposes but let's find a way to manipulate a properly-typed equivalent instead.

This comment also applies elsewhere.

Reply:
@postamar Let me think on this a bit.  My initial reaction is that we could add these methods to the operation interface (e.g. `requiresCatalog()`, `requiresConfig`()) and let Micronaut inject the configured operation and check if it supports it.  I believe that this will work, but I need to poke around a bit before declaring victory.

Reply:
I believe that this will work.  I will push up the fix to this branch so you can take a look at it.

Reply:
It's much more readable now, thank you.